url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/RQLibrary/Encoder.java

```diff
@@ -0,0 +1,2552 @@
+			    }
+			    
+			    // node2 already in graph?
+			    if ( graph.keySet().contains(node2) ) { // it is
+			    
+				// then lets add node 1 to its neighbours
+				graph.get(node2).add(node1);
+			    } else { // it isn't
+			    
+				// allocate memory for its neighbours
+				Set<Integer> edges = new HashSet<Integer>();
+				
+				// add node 1 to its neighbours
+				edges.add(node1);
+				
+				// finally, add node 2 to the graph along with its neighbours
+				graph.put(node2, edges);
+			    }
+			} else
+			    continue;
+		    }
+		    
+		    /*
+		     * the graph is complete, now we must find the maximum size component
+		     */
+		    
+		    // have we found the maximum size component yet?
+		    boolean found = false;
+		    
+		    // set of visited nodes
+		    Set<Integer> visited = null;
+		    
+		    /*
+		     * TODO Optimizacao: - ja procurei, e ha algoritmos optimizados para achar connected components e so depois ver qual o maior...
+		     */
+		    
+		    // what is the size of the largest component we've already found
+		    int maximumSize = 0;
+		    
+		    // the maximum size component
+		    Set<Integer> greatestComponent = null; // TODO testar tempos com isto
+		    
+		    // which nodes have already been used (either in visited or in toVisit)
+		    Set<Integer> used = new HashSet<Integer>();
+		    
+		    // iterates the nodes in the graph
+		    Iterator<Map.Entry<Integer, Set<Integer>>> it = graph.entrySet().iterator();
+		    
+		    // let's iterate through the nodes in the graph, looking for the maximum
+		    // size component. we will be doing a breadth first seach // TODO optimize this with a better algorithm?
+		    while (it.hasNext() && !found) {
+			// get our initial node
+			Map.Entry<Integer, Set<Integer>> node = it.next();
+			int initialNode = node.getKey();
+			
+			// we can't have used it before!
+			if ( used.contains(initialNode) )
+			    continue;
+			
+			// what are the edges of our initial node?
+			Integer[] edges = (Integer[]) node.getValue().toArray(new Integer[1]);
+			
+			// allocate memory for the set of visited nodes
+			visited = new HashSet<Integer>();
+			
+			// the set of nodes we must still visit
+			List<Integer> toVisit = new LinkedList<Integer>();
+			
+			// add the initial node to the set of used and visited nodes
+			visited.add(initialNode);
+			used.add(initialNode);
+			
+			// add my edges to the set of nodes we must visit
+			// and also put them in the used set
+			for (Integer edge : edges) {
+			    toVisit.add(edge);
+			    used.add(edge);
+			}
+			
+			// start the search!
+			while (toVisit.size() != 0) {
+			    // the node we are visiting
+			    int no = toVisit.remove(0);
+			    
+			    // add node to visited set
+			    visited.add(no);
+			    
+			    // queue edges to be visited (if they haven't been already
+			    for (Integer edge : graph.get(no))
+				if ( !visited.contains(edge) )
+				    toVisit.add(edge);
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1446:1457, 2333:2344.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.